src/Data/Text/Encoding.hs

```diff
@@ -391,7 +391,37 @@ decodeUtf8Lenient = decodeUtf8With lenientDecode
 --
 -- @since 1.1.0.0
 encodeUtf8Builder :: Text -> B.Builder
-encodeUtf8Builder = encodeUtf8BuilderEscaped (BP.liftFixedToBounded BP.word8)
+encodeUtf8Builder =
+    -- manual eta-expansion to ensure inlining works as expected
+    \txt -> B.builder (step txt)
+  where
+    step txt@(Text arr off len) !k br@(B.BufferRange op ope)
+      -- Ensure that the common case is not recursive and therefore yields
+      -- better code.
+      | op' <= ope = do
+          unsafeSTToIO $ A.copyToPointer arr off op len
+          k (B.BufferRange op' ope)
+      | otherwise = textCopyStep txt k br
+      where
+        op' = op `plusPtr` len
+{-# INLINE encodeUtf8Builder #-}
+
+textCopyStep :: Text -> B.BuildStep a -> B.BuildStep a
+textCopyStep (Text arr off len) k =
+    go off (off + len)
+  where
+    go !ip !ipe (B.BufferRange op ope)
+      | inpRemaining <= outRemaining = do
+          unsafeSTToIO $ A.copyToPointer arr ip op inpRemaining
+          let !br = B.BufferRange (op `plusPtr` inpRemaining) ope
+          k br
+      | otherwise = do
+          unsafeSTToIO $ A.copyToPointer arr ip op outRemaining
+          let !ip' = ip + outRemaining
+          return $ B.bufferFull 1 ope (go ip' ipe)
```

Comment:
It could be that I terribly misunderstand how `bufferFull` works, but why do we pass `1` instead of `inpRemaining - outRemaining`? We could have avoided any recursive functions at all, no?

Reply:
I think it signals „I need at least one byte to make progress“. As stated in the description this is copied more or less from the ByteString and ShortByteString cases.

Reply:
Comment on bufferFull arg says 
„Minimal size of next 'BufferRange'.“

Reply:
Yeah, I have the same question about implementations in `bytestring` :)

Reply:
So to me it looks like conservative default. If you’d pass the the length of the remaining string you’d put assumptions on the upper bound of the size of the buffer you are writing to while we really want the driver to be in charge of the buffer alone. I could def. see a compromise of passing upper and lower bound together so that the driver can decide how much to allocate more precisely.